packages/core/mesh/teleport/src/testing/test-extension-with-streams.ts

```diff
@@ -0,0 +1,233 @@
+//
+// Copyright 2022 DXOS.org
+//
+
+import assert from 'node:assert';
+import { randomBytes } from 'node:crypto';
+import { Duplex } from 'node:stream';
+
+import { Trigger } from '@dxos/async';
+import { log } from '@dxos/log';
+import { schema } from '@dxos/protocols';
+import { TestServiceWithStreams } from '@dxos/protocols/proto/example/testing/rpc';
+import { createProtoRpcPeer, ProtoRpcPeer } from '@dxos/rpc';
+
+import { ExtensionContext, TeleportExtension } from '../teleport';
+
+interface TestExtensionWithStreamsCallbacks {
+  onOpen?: () => Promise<void>;
+  onClose?: () => Promise<void>;
+}
+
+export class TestExtensionWithStreams implements TeleportExtension {
+  public readonly open = new Trigger();
+  public readonly closed = new Trigger();
+
+  public extensionContext: ExtensionContext | undefined;
+  private _rpc!: ProtoRpcPeer<{ TestServiceWithStreams: TestServiceWithStreams }>;
+
+  private readonly _streams = new Map<string, TestStream>();
+
+  constructor(public readonly callbacks: TestExtensionWithStreamsCallbacks = {}) {}
+
+  get remotePeerId() {
+    return this.extensionContext?.remotePeerId;
+  }
+
+  private _loadStream(streamTag: string, interval = 5, chunkSize = 2048) {
+    assert(!this._streams.has(streamTag), `Stream ${streamTag} already exists.`);
+
+    const networkStream = this.extensionContext!.createStream(streamTag, {
+      contentType: 'application/x-test-stream',
+    });
+
+    const streamEntry: TestStream = {
+      networkStream,
+      bytesSent: 0,
+      bytesReceived: 0,
+      sendErrors: 0,
+      receiveErrors: 0,
+      startTimestamp: Date.now(),
+    };
+
+    streamEntry.timer = setInterval(() => {
+      const chunk = randomBytes(chunkSize);
+      networkStream.write(chunk, 'binary', (err) => {
```

Comment:
`write` method returns true/false based on it having free space in the buffer for more chunks https://nodejs.org/api/stream.html#writablewritechunk-encoding-callback

you can add backpressure using that, though i'm not sure if it will work correctly in this test as it  needs to be correctly hooked up all the way to the transort